### PR TITLE
AvailabilityZone show on taskmetadata endpoint

### PR DESCRIPTION
**Pull request by cyastella**
### Summary

Expose availabilityZone to task metadata endpoint 
### Implementation details

Pass the availabilityZone from container instance level to agent metadata and expose availabilityZone from agent metadata to task response struct. 
### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: misc/v3-task-endpoint-validator/v3-task-endpoint-validator.go [hunk 1]
```diff
@@ -184,7 +185,7 @@ func verifyTaskMetadataResponse(taskMetadataRawMsg json.RawMessage) error {
 		"KnownStatus":   "RUNNING",
 	}
 
-	taskExpectedFieldNotEmptyArray := []string{"TaskARN", "Family", "Revision", "PullStartedAt", "PullStoppedAt", "Containers"}
+	taskExpectedFieldNotEmptyArray := []string{"TaskARN", "Family", "Revision", "PullStartedAt", "PullStoppedAt", "Containers","AvailabilityZone"}
```
Comment (linkar-ec2) [inline misc/v3-task-endpoint-validator/v3-task-endpoint-validator.go]: nit: add space

File: agent/api/ecsclient/client_test.go [hunk 2]
```diff
@@ -375,13 +376,16 @@ func TestReRegisterContainerInstance(t *testing.T) {
...
 
-	arn, err := client.RegisterContainerInstance("arn:test", capabilities, containerInstanceTags)
+	arn, availabilityzone,err := client.RegisterContainerInstance("arn:test", capabilities, containerInstanceTags)
 	if err != nil {
 		t.Errorf("Should not be an error: %v", err)
 	}
 	if arn != "registerArn" {
 		t.Errorf("Wrong arn: %v", arn)
 	}
+	if availabilityzone != "us-west-2b"{
```
[the remark this answers is not in the document]
Comment (petderek) [reply agent/api/ecsclient/client_test.go]: +1 This test uses a different style of test assertions -- you can migrate it if you want :)

[the remark this answers is not in the document]
Comment (cyastella) [reply agent/api/ecsclient/client_test.go]: good point, will refactor the test style.

**Comment by cyastella** [comment]
@linkar-ec2   v3 handler shares the same task metadata response format with v2 handler, so to keep consistent, we add availabilityZone to both places.

File: agent/api/ecsclient/client.go [hunk 3]
```diff
@@ -164,18 +164,30 @@ func (client *APIECSClient) registerContainerInstance(clusterRef string, contain
...
 
 	registerRequest.TotalResources = resources
 	resp, err := client.standardClient.RegisterContainerInstance(&registerRequest)
 	if err != nil {
 		seelog.Errorf("Unable to register as a container instance with ECS: %v", err)
-		return "", err
+		return "","", err
+	}
+
+	var availabilityzone_name = "ecs.availability-zone"
```
[the remark this answers is not in the document]
Comment (petderek) [reply agent/api/ecsclient/client.go]: Also, make this a const and put it at the top of the file

File: misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go [hunk 4]
```diff
@@ -41,6 +41,7 @@ type TaskResponse struct {
 	PullStartedAt      *time.Time          `json:"PullStartedAt,omitempty"`
 	PullStoppedAt      *time.Time          `json:"PullStoppedAt,omitempty"`
 	ExecutionStoppedAt *time.Time          `json:"ExecutionStoppedAt,omitempty"`
+	//AvailabilityZone   string              `json:"AvailabilityZone"`
```
Comment (yunhee-l) [inline misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go]: why is this commented out?

File: agent/app/agent_test.go [hunk 5]
```diff
@@ -103,10 +104,11 @@ func TestDoStartMinimumSupportedDockerVersionTerminal(t *testing.T) {
 		stateManagerFactory:   stateManagerFactory,
 		saveableOptionFactory: saveableOptionFactory,
 	}
-
+	assert.Equal(t, "", agent.availabilityzone)
```
Comment (petderek) [inline agent/app/agent_test.go]: I don't think we need this here. The test isn't doing anything AZ-specific

File: agent/app/agent_test.go [hunk 6]
```diff
@@ -103,10 +104,11 @@ func TestDoStartMinimumSupportedDockerVersionTerminal(t *testing.T) {
...
 	exitCode := agent.doStart(eventstream.NewEventStream("events", ctx),
 		credentialsManager, state, imageManager, client)
 	assert.Equal(t, exitcodes.ExitTerminal, exitCode)
+	assert.Equal(t,"", agent.availabilityzone)
```
Comment (petderek) [inline agent/app/agent_test.go]: I don't think we need this here either.

File: misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go (hunk 4 above)
Comment (petderek) [inline misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go]: Why is this commented out?

File: agent/app/agent_test.go [hunk 7]
```diff
@@ -763,10 +785,11 @@ func TestRegisterContainerInstanceWhenContainerInstanceARNIsNotSetHappyPath(t *t
 		credentialProvider: aws_credentials.NewCredentials(mockCredentialsProvider),
 		mobyPlugins:        mockMobyPlugins,
 	}
-
+	assert.Equal(t, "", agent.availabilityzone)
```
Comment (petderek) [inline agent/app/agent_test.go]: I don't think we need to check for emptiness before running the test.

File: agent/app/agent_test.go (hunk 6 above)
Comment (cyastella) [reply agent/app/agent_test.go]: will remove it.

File: misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go (hunk 4 above)
Comment (cyastella) [reply misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go]: will update it, haven't add it to windows yet.

Comment (cyastella) [reply misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go]: will update it, haven't add it to windows yet.

File: agent/api/ecsclient/client.go [hunk 8]
```diff
@@ -164,18 +164,30 @@ func (client *APIECSClient) registerContainerInstance(clusterRef string, contain
...
+			if aws.StringValue(attr.Name) == availabilityzone_name {
+				availabilityzone = aws.StringValue(attr.Value)
+			}
+		}
 	}
+
 	seelog.Info("Registered container instance with cluster!")
 	err = validateRegisteredAttributes(registerRequest.Attributes, resp.ContainerInstance.Attributes)
-	return aws.StringValue(resp.ContainerInstance.ContainerInstanceArn), err
+	return aws.StringValue(resp.ContainerInstance.ContainerInstanceArn), availabilityzone,err
```
Comment (haikuoliu) [inline agent/api/ecsclient/client.go]: nit: space, this happens a lot, have you configured `gofmt` on your editor? It will automatically adjust spaces.

File: agent/api/ecsclient/client.go [hunk 9]
```diff
@@ -164,18 +164,30 @@ func (client *APIECSClient) registerContainerInstance(clusterRef string, contain
...
+		return "","", err
+	}
+
+	var availabilityzone_name = "ecs.availability-zone"
+	var availabilityzone = ""
+	attrs := resp.ContainerInstance.Attributes
+	if(resp != nil){
+		for _, attr := range attrs {
+			if aws.StringValue(attr.Name) == availabilityzone_name {
+				availabilityzone = aws.StringValue(attr.Value)
```
Comment (haikuoliu) [inline agent/api/ecsclient/client.go]: You can add a break here

File: agent/handlers/v2/response.go [hunk 10]
```diff
@@ -88,6 +90,7 @@ func NewTaskResponse(taskARN string,
 		Revision:      task.Version,
 		DesiredStatus: task.GetDesiredStatus().String(),
 		KnownStatus:   task.GetKnownStatus().String(),
+		AvailabilityZone: az,
```
Comment (haikuoliu) [inline agent/handlers/v2/response.go]: please `gofmt` this

File: agent/app/agent.go [hunk 11]
```diff
@@ -321,15 +322,15 @@ func (agent *ecsAgent) newTaskEngine(containerChangeEventStream *eventstream.Eve
...
+	var previousCluster, previousEC2InstanceID, previousContainerInstanceArn, previousAZ string
 	previousTaskEngine := engine.NewTaskEngine(agent.cfg, agent.dockerClient,
 		credentialsManager, containerChangeEventStream, imageManager, state,
 		agent.metadataManager, agent.resourceFields)
 
 	// previousStateManager is used to verify that our current runtime configuration is
 	// compatible with our past configuration as reflected by our state-file
 	previousStateManager, err := agent.newStateManager(previousTaskEngine, &previousCluster,
-		&previousContainerInstanceArn, &previousEC2InstanceID)
+		&previousContainerInstanceArn, &previousEC2InstanceID, &previousAZ)
```
Comment (haikuoliu) [inline agent/app/agent.go]: Do you need to bump `ECSDataVersion`?

File: agent/api/ecsclient/client.go [hunk 12]
```diff
@@ -164,18 +165,30 @@ func (client *APIECSClient) registerContainerInstance(clusterRef string, contain
...
 	if err != nil {
 		seelog.Errorf("Unable to register as a container instance with ECS: %v", err)
-		return "", err
+		return "", "", err
+	}
+
+	var availabilityzone = ""
+	attrs := resp.ContainerInstance.Attributes
+	if resp != nil {
+		for _, attr := range attrs {
```
Comment (sharanyad) [inline agent/api/ecsclient/client.go]: nit:
this can be `for _, attr := range resp.ContainerInstance.Attributes {`

File: agent/app/agent.go [hunk 13]
```diff
@@ -427,6 +429,7 @@ func (agent *ecsAgent) newStateManager(
 		agent.saveableOptionFactory.AddSaveable("Cluster", cluster),
 		// This is for making testing easier as we can mock this
 		agent.saveableOptionFactory.AddSaveable("EC2InstanceID", savedInstanceID),
+		agent.saveableOptionFactory.AddSaveable("availabilityzone", availabilityzone),
```
Comment (sharanyad) [inline agent/app/agent.go]: nit: `AvailabilityZone` should be the key

File: agent/app/agent.go [hunk 14]
```diff
@@ -107,6 +107,7 @@ type ecsAgent struct {
 	terminationHandler    sighandlers.TerminationHandler
 	mobyPlugins           mobypkgwrapper.Plugins
 	resourceFields        *taskresource.ResourceFields
+	availabilityzone      string
```
Comment (sharanyad) [inline agent/app/agent.go]: nit: `availabilityZone`

File: misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go [hunk 15]
```diff
@@ -36,6 +36,7 @@ type TaskResponse struct {
 	Revision           string              `json:"Revision"`
 	DesiredStatus      string              `json:"DesiredStatus,omitempty"`
 	KnownStatus        string              `json:"KnownStatus"`
+	AvailabilityZone   string              `json:"AvailabilityZone"`
```
Comment (sharanyad) [inline misc/v3-task-endpoint-validator-windows/v3-task-endpoint-validator-windows.go]: I believe similar changes are needed in [this](https://github.com/aws/amazon-ecs-agent/blob/master/misc/taskmetadata-validator/taskmetadata-validator.go) file.

File: agent/api/ecsclient/client.go [hunk 16]
```diff
@@ -41,6 +41,7 @@ const (
 	pollEndpointCacheSize = 1
 	pollEndpointCacheTTL  = 20 * time.Minute
 	roundtripTimeout      = 5 * time.Second
+	availabilityZoneName  = "ecs.availability-zone"
```
Comment (sharanyad) [inline agent/api/ecsclient/client.go]: how about naming this `azAttrName`?

File: agent/api/task/task.go [hunk 17]
```diff
@@ -94,6 +94,8 @@ type TaskOverrides struct{}
 type Task struct {
 	// Arn is the unique identifier for the task
 	Arn string
+	//AvailabilityZone is the zone for task locates in
+	AvailabilityZone string
```
Comment (sharanyad) [inline agent/api/task/task.go]: why is this needed? as I understand, we use agent's availabilityZone field for v2/v3 responses, right?
